testrepository/commands/run.py

```diff
@@ -146,19 +151,34 @@ class run(Command):
         optparse.Option("--isolated", action="store_true",
             default=False,
             help="Run each test id in a separate test runner."),
+        optparse.Option("-p", "--profiles",
+            help="Comma separated list of profiles that may be present in the "
+                 "stream. By default the list_profiles hook in testr.conf is "
+                 "queried.",
+            default=""),
         ]
     args = [StringArgument('testfilters', 0, None), DoubledashArgument(),
         StringArgument('testargs', 0, None)]
     # Can be assigned to to inject a custom command factory.
     command_factory = TestCommand
 
-    def _find_failing(self, repo):
+    def _find_failing(self, repo, profiles):
         run = repo.get_failing()
         case = run.get_test()
-        ids = []
+        ids = {}
         def gather_errors(test_dict):
```

Comment:
I suggest extracting this to a top-level private function, and giving it a docstring & maybe tests.